File: Sources/Client/Camera.cpp
#include "Camera.h"
#include <cstdlib>
#include <cmath>

Camera& Camera::Get()
{
	static Camera instance;
	return instance;
}

void Camera::Initialize()
{
	m_mode = CameraMode::FollowPlayer;
	m_attachedEntityId = 0;

	m_viewX = 0;
	m_viewY = 0;
	m_exactX = 0.0f;
	m_exactY = 0.0f;
	m_targetX = 0;
	m_targetY = 0;

	m_bSmoothFollow = false;
	m_followSpeed = 0.1f;

	// Default viewport (will be set properly by game init)
	m_viewWidth = 800;
	m_viewHeight = 600;
	m_tileWidth = 32;
	m_tileHeight = 32;
	m_viewWidthTiles = m_viewWidth / m_tileWidth + 2;
	m_viewHeightTiles = m_viewHeight / m_tileHeight + 2;

	m_bShaking = false;
	m_shakeDegree = 0;
	m_shakeDuration = 0.0f;
	m_shakeTimer = 0.0f;
	m_shakeOffsetX = 0;
	m_shakeOffsetY = 0;

	m_worldMinX = 0;
	m_worldMinY = 0;
	m_worldMaxX = 0;
	m_worldMaxY = 0;
	m_bClampToBounds = false;
}

void Camera::Shutdown()
{
	// Nothing to clean up
}

void Camera::SetViewportSize(int width, int height)
{
	m_viewWidth = width;
	m_viewHeight = height;
	m_viewWidthTiles = width / m_tileWidth + 2;
	m_viewHeightTiles = height / m_tileHeight + 2;
}

void Camera::SetTileSize(int tileWidth, int tileHeight)
{
	m_tileWidth = tileWidth;
	m_tileHeight = tileHeight;
	m_viewWidthTiles = m_viewWidth / m_tileWidth + 2;
	m_viewHeightTiles = m_viewHeight / m_tileHeight + 2;
}

void Camera::AttachToPlayer()
{
	m_mode = CameraMode::FollowPlayer;
	m_attachedEntityId = 0;
}

void Camera::AttachToEntity(uint32_t entityId)
{
	m_mode = CameraMode::FollowEntity;
	m_attachedEntityId = entityId;
}

void Camera::Detach()
{
	m_mode = CameraMode::FreeCamera;
}

void Camera::SetFixedPosition(int worldX, int worldY)
{
	m_mode = CameraMode::Fixed;
	m_viewX = worldX - m_viewWidthTiles / 2;
	m_viewY = worldY - m_viewHeightTiles / 2;
	m_exactX = static_cast<float>(m_viewX);
	m_exactY = static_cast<float>(m_viewY);
	ClampToBounds();
}

int Camera::GetViewX() const
{
	return m_viewX + m_shakeOffsetX;
}

int Camera::GetViewY() const
{
	return m_viewY + m_shakeOffsetY;
}

int Camera::GetCenterX() const
{
	return m_viewX + m_viewWidthTiles / 2;
}

int Camera::GetCenterY() const
{
	return m_viewY + m_viewHeightTiles / 2;
}

void Camera::GetViewPosition(int& x, int& y) const
{
	x = GetViewX();
	y = GetViewY();
}

void Camera::SetPosition(int tileX, int tileY)
{
	if (m_mode == CameraMode::FreeCamera || m_mode == CameraMode::Fixed)
	{
		m_viewX = tileX;
		m_viewY = tileY;
		m_exactX = static_cast<float>(m_viewX);
		m_exactY = static_cast<float>(m_viewY);
		ClampToBounds();
	}
}

void Camera::SetViewPosition(int tileX, int tileY)
{
	m_viewX = tileX;
	m_viewY = tileY;
	m_exactX = static_cast<float>(m_viewX);
	m_exactY = static_cast<float>(m_viewY);
	m_targetX = tileX;
	m_targetY = tileY;
}

void Camera::Move(int deltaTileX, int deltaTileY)
{
	if (m_mode == CameraMode::FreeCamera)
	{
		m_viewX += deltaTileX;
		m_viewY += deltaTileY;
		m_exactX = static_cast<float>(m_viewX);
		m_exactY = static_cast<float>(m_viewY);
		ClampToBounds();
	}
}

void Camera::SetSmoothFollow(bool enabled)
{
	m_bSmoothFollow = enabled;
}

void Camera::SetFollowSpeed(float speed)
{
	if (speed < 0.01f) speed = 0.01f;
	if (speed > 1.0f) speed = 1.0f;
	m_followSpeed = speed;
}

void Camera::Shake(int intensity, float duration)
{
	m_bShaking = true;
	m_shakeDegree = intensity;
	m_shakeDuration = duration;
	m_shakeTimer = 0.0f;
}

void Camera::StopShake()
{
	m_bShaking = false;
	m_shakeDegree = 0;
	m_shakeOffsetX = 0;
	m_shakeOffsetY = 0;
}

void Camera::WorldToScreen(int worldX, int worldY, int& screenX, int& screenY) const
{
	screenX = (worldX - GetViewX()) * m_tileWidth;
	screenY = (worldY - GetViewY()) * m_tileHeight;
}

void Camera::ScreenToWorld(int screenX, int screenY, int& worldX, int& worldY) const
{
	worldX = GetViewX() + screenX / m_tileWidth;
	worldY = GetViewY() + screenY / m_tileHeight;
}

bool Camera::IsWorldPosVisible(int worldX, int worldY) const
{
	int left, top, right, bottom;
	GetVisibleTileBounds(left, top, right, bottom);

	return worldX >= left && worldX <= right &&
		worldY >= top && worldY <= bottom;
}

bool Camera::IsTileVisible(int tileX, int tileY) const
{
	return IsWorldPosVisible(tileX, tileY);
}

void Camera::GetVisibleTileBounds(int& left, int& top, int& right, int& bottom) const
{
	left = GetViewX();
	top = GetViewY();
	right = left + m_viewWidthTiles;
	bottom = top + m_viewHeightTiles;
}

void Camera::SetWorldBounds(int minX, int minY, int maxX, int maxY)
{
	m_worldMinX = minX;
	m_worldMinY = minY;
	m_worldMaxX = maxX;
	m_worldMaxY = maxY;
	m_bClampToBounds = true;
	ClampToBounds();
}

void Camera::ClearWorldBounds()
{
	m_bClampToBounds = false;
}

void Camera::Update(double deltaTime)
{
	UpdateFollow(deltaTime);
	UpdateShake(deltaTime);
}

void Camera::UpdateFollow(double deltaTime)
{
	// For now, camera following is handled externally by CGame
	// This will be used when we fully integrate EntityManager/PlayerManager

	if (m_mode == CameraMode::FreeCamera || m_mode == CameraMode::Fixed)
	{
		return;
	}

	// If smooth follow is enabled and we have a target
	if (m_bSmoothFollow && (m_targetX != m_viewX || m_targetY != m_viewY))
	{
		float t = 1.0f - powf(1.0f - m_followSpeed, static_cast<float>(deltaTime) * 60.0f);
		m_exactX += (m_targetX - m_exactX) * t;
		m_exactY += (m_targetY - m_exactY) * t;
		m_viewX = static_cast<int>(m_exactX);
		m_viewY = static_cast<int>(m_exactY);
		ClampToBounds();
	}
}

void Camera::UpdateShake(double deltaTime)
{
	if (!m_bShaking) return;

	m_shakeTimer += static_cast<float>(deltaTime);

	if (m_shakeTimer >= m_shakeDuration)
	{
		StopShake();
		return;
	}

	// Decrease intensity over time
	float remaining = 1.0f - (m_shakeTimer / m_shakeDuration);
	int currentIntensity = static_cast<int>(m_shakeDegree * remaining);

	m_shakeOffsetX = GenerateShakeOffset(currentIntensity);
	m_shakeOffsetY = GenerateShakeOffset(currentIntensity);
}

void Camera::ClampToBounds()
{
	if (!m_bClampToBounds) return;

	if (m_viewX < m_worldMinX) m_viewX = m_worldMinX;
	if (m_viewY < m_worldMinY) m_viewY = m_worldMinY;

	int maxViewX = m_worldMaxX - m_viewWidthTiles;
	int maxViewY = m_worldMaxY - m_viewHeightTiles;

	if (m_viewX > maxViewX) m_viewX = maxViewX;
	if (m_viewY > maxViewY) m_viewY = maxViewY;

	m_exactX = static_cast<float>(m_viewX);
	m_exactY = static_cast<float>(m_viewY);
}

int Camera::GenerateShakeOffset(int intensity) const
{
	if (intensity <= 0) return 0;
	return (rand() % (intensity * 2 + 1)) - intensity;
}

File: engine/src/core/Camera.ts
/**
 * Camera — Viewport camera singleton.
 * Mirrors C++ Camera from Sources/Client/Camera.h/.cpp.
 *
 * Manages the view position (in tile coordinates), smooth following,
 * screen shake, coordinate conversion, and world bounds clamping.
 */

// ── CameraMode enum ──────────────────────────────────────────────

export enum CameraMode {
	FollowPlayer = 0,
	FollowEntity,
	FreeCamera,
	Fixed,
}

// ── Camera class ─────────────────────────────────────────────────

export class Camera {
	private static s_instance: Camera | null = null;

	static get(): Camera {
		if (!Camera.s_instance) {
			Camera.s_instance = new Camera();
		}
		return Camera.s_instance;
	}

	// ── Mode ──────────────────────────────────────────────────────

	private m_mode: CameraMode = CameraMode.FollowPlayer;
	private m_attachedEntityId = 0;

	// ── Position (in tiles) ───────────────────────────────────────

	private m_viewX = 0;   // top-left tile X
	private m_viewY = 0;   // top-left tile Y
	private m_exactX = 0;  // smooth sub-tile X
	private m_exactY = 0;  // smooth sub-tile Y
	private m_targetX = 0;
	private m_targetY = 0;

	// ── Viewport dimensions ───────────────────────────────────────

	private m_viewWidth = 640;
	private m_viewHeight = 480;
	private m_tileWidth = 32;
	private m_tileHeight = 32;
	private m_viewWidthTiles = 20;
	private m_viewHeightTiles = 15;

	// ── Smooth follow ─────────────────────────────────────────────

	private m_bSmoothFollow = true;
	private m_followSpeed = 8.0;

	// ── Shake ─────────────────────────────────────────────────────

	private m_bShaking = false;
	private m_shakeDegree = 0;
	private m_shakeDuration = 0;
	private m_shakeTimer = 0;
	private m_shakeOffsetX = 0;
	private m_shakeOffsetY = 0;

	// ── World bounds ──────────────────────────────────────────────

	private m_worldMinX = 0;
	private m_worldMinY = 0;
	private m_worldMaxX = 0;
	private m_worldMaxY = 0;
	private m_bClampToBounds = false;

	// ── Lifecycle ─────────────────────────────────────────────────

	initialize(): void {
		this.m_mode = CameraMode.FollowPlayer;
		this.m_viewX = 0;
		this.m_viewY = 0;
		this.m_exactX = 0;
		this.m_exactY = 0;
		this.m_bShaking = false;
	}

	shutdown(): void {
		// Nothing to release in TS
	}

	// ── Viewport setup ────────────────────────────────────────────

	setViewportSize(width: number, height: number): void {
		this.m_viewWidth = width;
		this.m_viewHeight = height;
		this.m_viewWidthTiles = Math.ceil(width / this.m_tileWidth);
		this.m_viewHeightTiles = Math.ceil(height / this.m_tileHeight);
	}

	setTileSize(tileWidth: number, tileHeight: number): void {
		this.m_tileWidth = tileWidth;
		this.m_tileHeight = tileHeight;
		this.m_viewWidthTiles = Math.ceil(this.m_viewWidth / tileWidth);
		this.m_viewHeightTiles = Math.ceil(this.m_viewHeight / tileHeight);
	}

	// ── Attachment ────────────────────────────────────────────────

	attachToPlayer(): void {
		this.m_mode = CameraMode.FollowPlayer;
		this.m_attachedEntityId = 0;
	}

	attachToEntity(id: number): void {
		this.m_mode = CameraMode.FollowEntity;
		this.m_attachedEntityId = id;
	}

	detach(): void {
		this.m_mode = CameraMode.FreeCamera;
		this.m_attachedEntityId = 0;
	}

	setFixedPosition(worldX: number, worldY: number): void {
		this.m_mode = CameraMode.Fixed;
		this.m_targetX = worldX;
		this.m_targetY = worldY;
		this.m_exactX = worldX;
		this.m_exactY = worldY;
	}

	getMode(): CameraMode { return this.m_mode; }
	getAttachedEntityId(): number { return this.m_attachedEntityId; }
	isAttachedToPlayer(): boolean { return this.m_mode === CameraMode.FollowPlayer; }

	// ── Position accessors ────────────────────────────────────────

	/** Top-left tile X of the viewport. */
	getViewX(): number { return this.m_viewX; }
	/** Top-left tile Y of the viewport. */
	getViewY(): number { return this.m_viewY; }
	/** Center tile X of the viewport. */
	getCenterX(): number { return this.m_viewX + Math.floor(this.m_viewWidthTiles / 2); }
	/** Center tile Y of the viewport. */
	getCenterY(): number { return this.m_viewY + Math.floor(this.m_viewHeightTiles / 2); }

	getViewPosition(): [number, number] { return [this.m_viewX, this.m_viewY]; }

	/** Set the target position for the camera (in tiles). */
	setPosition(tileX: number, tileY: number): void {
		this.m_targetX = tileX;
		this.m_targetY = tileY;
		if (!this.m_bSmoothFollow) {
			this.m_exactX = tileX;
			this.m_exactY = tileY;
		}
	}

	/** Move target by delta tiles. */
	move(dx: number, dy: number): void {
		this.m_targetX += dx;
		this.m_targetY += dy;
	}

	/** Snap the view directly (bypassing smooth follow). */
	setViewPosition(tileX: number, tileY: number): void {
		this.m_exactX = tileX;
		this.m_exactY = tileY;
		this.m_targetX = tileX;
		this.m_targetY = tileY;
	}

	// ── Smooth follow ─────────────────────────────────────────────

	setSmoothFollow(enabled: boolean): void { this.m_bSmoothFollow = enabled; }
	setFollowSpeed(speed: number): void { this.m_followSpeed = speed; }
	isSmoothFollow(): boolean { return this.m_bSmoothFollow; }

	// ── Shake ─────────────────────────────────────────────────────

	shake(intensity: number, duration: number): void {
		this.m_bShaking = true;
		this.m_shakeDegree = intensity;
		this.m_shakeDuration = duration;
		this.m_shakeTimer = 0;
	}

	stopShake(): void {
		this.m_bShaking = false;
		this.m_shakeOffsetX = 0;
		this.m_shakeOffsetY = 0;
	}

	isShaking(): boolean { return this.m_bShaking; }
	getShakeOffsetX(): number { return this.m_shakeOffsetX; }
	getShakeOffsetY(): number { return this.m_shakeOffsetY; }
	getShakeDegree(): number { return this.m_shakeDegree; }

	// ── Coordinate conversion ─────────────────────────────────────

	/** Convert world pixel coordinates to screen coordinates. */
	worldToScreen(worldX: number, worldY: number): [number, number] {
		return [
			worldX - this.m_viewX * this.m_tileWidth + this.m_shakeOffsetX,
			worldY - this.m_viewY * this.m_tileHeight + this.m_shakeOffsetY,
		];
	}

	/** Convert screen coordinates to world pixel coordinates. */
	screenToWorld(screenX: number, screenY: number): [number, number] {
		return [
			screenX + this.m_viewX * this.m_tileWidth - this.m_shakeOffsetX,
			screenY + this.m_viewY * this.m_tileHeight - this.m_shakeOffsetY,
		];
	}

	// ── Visibility testing ────────────────────────────────────────

	isWorldPosVisible(worldX: number, worldY: number): boolean {
		const [sx, sy] = this.worldToScreen(worldX, worldY);
		return sx >= 0 && sx < this.m_viewWidth && sy >= 0 && sy < this.m_viewHeight;
	}

	isTileVisible(tileX: number, tileY: number): boolean {
		return tileX >= this.m_viewX && tileX < this.m_viewX + this.m_viewWidthTiles
			&& tileY >= this.m_viewY && tileY < this.m_viewY + this.m_viewHeightTiles;
	}

	getVisibleTileBounds(): { left: number; top: number; right: number; bottom: number } {
		return {
			left: this.m_viewX,
			top: this.m_viewY,
			right: this.m_viewX + this.m_viewWidthTiles,
			bottom: this.m_viewY + this.m_viewHeightTiles,
		};
	}

	// ── Viewport dimensions ───────────────────────────────────────

	getViewWidth(): number { return this.m_viewWidth; }
	getViewHeight(): number { return this.m_viewHeight; }
	getViewWidthTiles(): number { return this.m_viewWidthTiles; }
	getViewHeightTiles(): number { return this.m_viewHeightTiles; }
	getTileWidth(): number { return this.m_tileWidth; }
	getTileHeight(): number { return this.m_tileHeight; }

	// ── World bounds ──────────────────────────────────────────────

	setWorldBounds(minX: number, minY: number, maxX: number, maxY: number): void {
		this.m_worldMinX = minX;
		this.m_worldMinY = minY;
		this.m_worldMaxX = maxX;
		this.m_worldMaxY = maxY;
		this.m_bClampToBounds = true;
	}

	clearWorldBounds(): void {
		this.m_bClampToBounds = false;
	}

	// ── Per-frame update ──────────────────────────────────────────

	update(deltaTime: number): void {
		// Smooth follow towards target
		if (this.m_bSmoothFollow) {
			const speed = this.m_followSpeed * deltaTime;
			this.m_exactX += (this.m_targetX - this.m_exactX) * Math.min(1, speed);
			this.m_exactY += (this.m_targetY - this.m_exactY) * Math.min(1, speed);
		} else {
			this.m_exactX = this.m_targetX;
			this.m_exactY = this.m_targetY;
		}

		// Clamp to world bounds
		if (this.m_bClampToBounds) {
			this.m_exactX = Math.max(this.m_worldMinX, Math.min(this.m_worldMaxX - this.m_viewWidthTiles, this.m_exactX));
			this.m_exactY = Math.max(this.m_worldMinY, Math.min(this.m_worldMaxY - this.m_viewHeightTiles, this.m_exactY));
		}

		// Integer view position for rendering
		this.m_viewX = Math.floor(this.m_exactX);
		this.m_viewY = Math.floor(this.m_exactY);

		// Shake
		if (this.m_bShaking) {
			this.m_shakeTimer += deltaTime;
			if (this.m_shakeTimer >= this.m_shakeDuration) {
				this.stopShake();
			} else {
				const fade = 1.0 - this.m_shakeTimer / this.m_shakeDuration;
				const amp = this.m_shakeDegree * fade;
				this.m_shakeOffsetX = Math.round((Math.random() * 2 - 1) * amp);
				this.m_shakeOffsetY = Math.round((Math.random() * 2 - 1) * amp);
			}
		}
	}
}
